File: packages/pageflow/spec/editor/api/WidgetTypes-spec.js
describe('WidgetTypes', () => {
  describe('#register/#findAllByRole', () => {
    test(
      'allows getting a list of widget types with name and translation key',
      () => {
        var widgetTypes = new pageflow.WidgetTypes();

        widgetTypes.setup({
          navigation: [
            {
              name: 'fancy_bar',
              translationKey: 'pageflow.fancy_bar.widget_type_name'
            }
          ]
        });

        var widgetType = widgetTypes.findAllByRole('navigation')[0];

        expect(widgetType.name).toBe('fancy_bar');
        expect(widgetType.translationKey).toBe('pageflow.fancy_bar.widget_type_name');
      }
    );

    test('returns empty array for unknown role', () => {
      var widgetTypes = new pageflow.WidgetTypes();

      widgetTypes.setup({});

      expect(widgetTypes.findAllByRole('not_there')).toEqual([]);
    });
  });

  describe('#register/#findByName', () => {
    test('allows getting a widget type by name', () => {
      var widgetTypes = new pageflow.WidgetTypes();

      widgetTypes.setup({
        navigation: [
          {
            name: 'fancy_bar',
            translationKey: 'pageflow.fancy_bar.widget_type_name'
          }
        ]
      });

      var widgetType = widgetTypes.findByName('fancy_bar');

      expect(widgetType.name).toBe('fancy_bar');
    });
  });

  describe('#register/#setup/#findByName', () => {
    test(
      'allows supplying a configuration editor view for a widget type',
      () => {
        var widgetTypes = new pageflow.WidgetTypes();
        var viewConstructor = sinon.spy();

        widgetTypes.register('fancy_bar', {
          configurationEditorView: viewConstructor
        });
        widgetTypes.setup({
          navigation: [
            {
              name: 'fancy_bar',
              translationKey: 'pageflow.fancy_bar.widget_type_name'
            }
          ]
        });

        var widgetType = widgetTypes.findByName('fancy_bar');
        widgetType.createConfigurationEditorView({some: 'option'});

        expect(viewConstructor).to.have.been.calledWith(sinon.match({some: 'option'}));
      }
    );
  });
});

File: spec/pageflow/widget_types_spec.rb
require 'spec_helper'

module Pageflow
  describe WidgetTypes do
    before do
      Pageflow.config.widget_types.clear
    end

    describe '#roles' do
      it 'returns uniq list of roles' do
        widget_types = WidgetTypes.new
        widget_types.register(TestWidgetType.new(name: 'some_navigation', roles: ['navigation']))
        widget_types.register(TestWidgetType.new(name: 'some_header',
                                                 roles: [
                                                   'navigation', 'header'
                                                 ]))

        expect(widget_types.roles).to eq(['navigation', 'header'])
      end
    end

    describe '#defaults_by_role' do
      it 'returns default widget type for role' do
        widget_types = WidgetTypes.new
        widget_types.register(TestWidgetType.new(name: 'some_navigation', roles: ['navigation']),
                              default: true)

        expect(widget_types.defaults_by_role['navigation'].name).to eq('some_navigation')
      end
    end
  end
end
